pkg/clusters/types/gke/builder.go

```diff
@@ -176,3 +194,67 @@ func (b *Builder) Build(ctx context.Context) (clusters.Cluster, error) {
 
 	return cluster, nil
 }
+
+// createCluster creates the GKE cluster asynchronously.
+func (b *Builder) createCluster(ctx context.Context, req *containerpb.CreateClusterRequest, mgrc *container.ClusterManagerClient, createdByID, authToken string) error {
+	// createSubnet is currently only available via gcloud CLI:
+	// https://github.com/googleapis/google-cloud-go/issues/7219
+	if b.createSubnet {
+		if err := b.createClusterUsingCLI(ctx, req, createdByID, authToken); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	_, err := mgrc.CreateCluster(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *Builder) createClusterUsingCLI(ctx context.Context, req *containerpb.CreateClusterRequest, createdByID, authToken string) error {
+	tokenFile, err := os.CreateTemp("", "gcloud-token-")
+	if err != nil {
+		return fmt.Errorf("failed to create a temporary file for gcloud token: %w", err)
+	}
+	defer func() {
+		_ = os.Remove(tokenFile.Name())
+	}()
+	if _, err := io.WriteString(tokenFile, authToken); err != nil {
+		return fmt.Errorf("failed to write a token to the temporary file: %w", err)
+	}
+
+	//nolint:gosec
+	cmd := exec.CommandContext(ctx, "gcloud", "container", "clusters", "create", req.Cluster.Name,
+		`--access-token-file`, tokenFile.Name(),
+		`--project`, b.project,
+		`--region`, b.location,
+		`--create-subnetwork`, ``,
+		`--enable-ip-alias`,
+		`--num-nodes`, `1`,
+		`--cluster-version`, req.Cluster.InitialClusterVersion,
+		`--addons`, `HorizontalPodAutoscaling`,
+		`--labels`, fmt.Sprintf(`%s=%s`, GKECreateLabel, createdByID),
+		`--async`,
+	)
+	stderr := &bytes.Buffer{}
+	cmd.Stderr = stderr
+
+	if err := cmd.Run(); err != nil {
+		fmt.Println(stderr.String())
+		return fmt.Errorf("failed to run gcloud CLI: %w", err)
+	}
+
+	return nil
+}
+
+func sanitizeCreatedByID(id string) string {
+	// Replace dots with dashes as dots are not allowed by the CLI:
+	// "It must only contain lowercase letters ([a-z]), numeric characters ([0-9]), underscores (_) and dashes (-)."
+	id = strings.ReplaceAll(id, ".", "-")
```

Comment:
Since "It must only contain lowercase letters ([a-z]), numeric characters ([0-9]), underscores (_) and dashes (-).", replacing only dots with dashes ensures us to not have invalid `CreatedByID`s? Shouldn't we also deal with the other limitations?

Reply:
Yup, good point :) I was so focused on my case that I didn't take into account different possibilities: f4267ef